core/src/main/scala/com/databricks/labs/remorph/discovery/SnowflakeTableDefinitions.scala

```diff
@@ -0,0 +1,123 @@
+   * @return The DataType object corresponding to the input string.
+   */
+  private def getDataType(dataTypeString: String): DataType = {
+    val inputString = CharStreams.fromString(dataTypeString)
+    val lexer = new SnowflakeLexer(inputString)
+    val tokenStream = new CommonTokenStream(lexer)
+    val parser = new SnowflakeParser(tokenStream)
+    val ctx = parser.dataType()
+    val dataTypeBuilder = new SnowflakeTypeBuilder
+    dataTypeBuilder.buildDataType(ctx)
+  }
+
+  private def getTableDefinitionQuery(catalogName: String): String = {
+    s"""WITH column_info AS (
+       |    SELECT
+       |        TABLE_CATALOG,
+       |        TABLE_SCHEMA,
+       |        TABLE_NAME,
+       |        LISTAGG(column_name || ':' ||
+       |            CASE
+       |                WHEN numeric_precision IS NOT NULL AND numeric_scale IS NOT NULL
+       |                THEN
+       |                    CONCAT(data_type, '(', numeric_precision, ',' , numeric_scale, ')')
+       |                WHEN LOWER(data_type) = 'text'
+       |                THEN
+       |                    CONCAT('varchar', '(', CHARACTER_MAXIMUM_LENGTH, ')')
+       |                ELSE data_type
+       |             END|| ':' || TO_BOOLEAN(CASE WHEN IS_NULLABLE = 'YES' THEN 'true' ELSE 'false' END),
+       |        '~') WITHIN GROUP (ORDER BY ordinal_position) AS Schema
+       |    FROM
+       |        ${catalogName}.INFORMATION_SCHEMA.COLUMNS
+       |    GROUP BY
+       |        TABLE_CATALOG, TABLE_SCHEMA, TABLE_NAME
+       |)
+       |SELECT
+       |    sft.TABLE_CATALOG,
+       |    sft.TABLE_SCHEMA,
+       |    sft.TABLE_NAME,
+       |    sfe.location,
+       |    sfe.file_format_name,
+       |    sfv.view_definition,
+       |    column_info.Schema as DERIVED_SCHEMA,
+       |    sft.BYTES
+       |FROM
+       |    column_info
+       |JOIN ${catalogName}.INFORMATION_SCHEMA.TABLES sft
+       |    ON column_info.TABLE_CATALOG = sft.TABLE_CATALOG
+       |    AND column_info.TABLE_SCHEMA = sft.TABLE_SCHEMA
+       |    AND column_info.TABLE_NAME = sft.TABLE_NAME
+       |LEFT JOIN ${catalogName}.INFORMATION_SCHEMA.VIEWS sfv
+       |    ON column_info.TABLE_CATALOG = sfv.TABLE_CATALOG
+       |    AND column_info.TABLE_SCHEMA = sfv.TABLE_SCHEMA
+       |    AND column_info.TABLE_NAME = sfv.TABLE_NAME
+       |LEFT JOIN ${catalogName}.INFORMATION_SCHEMA.EXTERNAL_TABLES sfe
+       |    ON column_info.TABLE_CATALOG = sfe.TABLE_CATALOG
+       |    AND column_info.TABLE_SCHEMA = sfe.TABLE_SCHEMA
+       |    AND column_info.TABLE_NAME = sfe.TABLE_NAME
+       |ORDER BY
+       |    sft.TABLE_CATALOG, sft.TABLE_SCHEMA, sft.TABLE_NAME;""".stripMargin
+  }
+
+  /**
+   * Retrieves the definitions of all tables in the Snowflake database.
+   *
+   * @return A sequence of TableDefinition objects representing the tables in the database.
+   */
+  def getTableDefinitions(catalogName: String): Seq[TableDefinition] = {
```

Comment:
this should be private method, and you need to list all catalogs and get all table definitions from all catalogs.

can you also add methods to retrieve all catalogs and method to retrieve all schemas

Reply:
done